### ISSUE TITLE
sqlite3 for Windows 10 Universal apps

### ISSUE THREAD

**Issue by mjracca** [ericsink/SQLitePCL.raw#44]
Hi,

I'm migrating my app to run on Windows 10 and I'm having issues with sqlite3.dll. 

I've downloaded the latest libraries from the official website and replace it and it's working now, at least on x86.

Any change you could update this libraries for x86 and x64 on NuGet?

Thanks

**Comment by ericsink** [comment]
Yeah, it sounds like I need to make a change here.

But I haven't even looked at building for Windows 10 yet.  I suspect that just getting the tooling setup will take me some time.

For now, can you tell me anything more specific about the issues you were having?

Thanks!

**Comment by mjracca** [comment]
**Unable to load DLL 'sqlite3': The specified module could not be found.**

```
An exception of type 'System.TypeInitializationException' occurred in mscorlib.dll but was not handled in user code
Additional information: The type initializer for 'SQLitePCL.raw' threw an exception.
{"The type initializer for 'SQLitePCL.raw' threw an exception."}
{"Unable to load DLL 'sqlite3': The specified module could not be found. (Exception from HRESULT: 0x8007007E)"}
The type initializer for 'SQLitePCL.raw' threw an exception.
at SQLitePCL.raw.sqlite3_open_v2(String filename, sqlite3& db, Int32 flags, String vfs)
```

Thanks!

**Comment by ericsink** [comment]
That error message looks exactly like the one that people get when they use SQLitePCL.raw without adding a reference to the Visual C++ runtime extension SDK.  With any of the RT flavors of Windows (Windows Store, WP81, etc), this happens all the time.

I'm not sure which nuget platform from SQLitePCL.raw got included by your Windows 10 app.  Plain net45?  One of the RT flavors?

When you say that you "downloaded the latest libraries from the official website", I assume you are referring to sqlite.org?  Can you tell me exactly which library you downloaded?

I suspect that you just needed to add a C++ runtime reference.  I further suspect that the reason that using a build from sqlite.org worked is because that build had the C runtime lib statically linked.  Which AFAIK is impossible for any of the RT flavors.

But all this conjecture comes from a position of ignorance about how all these issues might be changing with Windows 10.

**Comment by mjracca** [comment]
Eric, sorry for the lack of context!

I'm using Akavache in my project, which uses SQLitePCL.raw. 

In fact, the Visual C++ runtime extension SDK is actually missing from the project. MSFT Technical Evangelist told me that the package isn't ready yet. Do you think this will solve the issue once they have it?

The library I've downloaded is "sqlite-dll-win32-x86-3080900.zip" from https://www.sqlite.org/download.html

I've replaced sqlite3.dll  in "packages\SQLitePCL.raw_basic.0.7.0\build\native\sqlite3_dynamic\winrt81" with the new one and it worked.

**Comment by ericsink** [comment]
I'm pretty sure the sqlite DLL you grabbed from sqlite.org was probably compiled for plain old win32, not with the newer toolchain.  I would not expect that to work.  But it did.  That's interesting.

I'll need to dig into this further.

And since you're on Akavache, I'll just tag @paulcbetts here to make sure he sees this.

**Comment by paulcbetts** [comment]
So while I've got you here, why don't we statically link the CRT? You mentioned it fails on RT - why?

**Comment by ericsink** [comment]
I would prefer to statically link it, but I think it's just not a supported
operation.

I can't find an authoritative link at the moment, but here's a stack
overflow post that says the kind of things I tend to see on this topic:

http://stackoverflow.com/questions/24727423/how-to-statically-link-to-msvcp120-dll-in-vs2013

IIRC, the last time I tried to make this work by directly manipulating the
command line flags, it gave me a compiler error.

--
E

**Comment by paulcbetts** [comment]
This whole SO post is like one big "Citation Needed" :P

**Comment by ericsink** [comment]
Yeah.  This is a bit more authoritative, although written in a confusing manner:

https://msdn.microsoft.com/en-us/library/hh972425.aspx

**Comment by theofanis** [comment]
I can confirm that using the dll from http://www.sqlite.org/2015/sqlite-dll-win32-x86-3081002.zip works for my WPF app using Akavache.

**Comment by dominik-weber** [comment]
Any updates on this issue so far? What I found out:
* Running on Windows 10 Desktop seems to work fine if a reference to VC++ 2013 libs is added
* Running on Windows 10 Mobile emulator breaks

I narrowed the problem down to the provided sqlite3.dll, build\native\sqlite3_dynamic\winrt81\x86\sqlite3.dll

If I remove this dll and add a reference to the sqlite3.dll provided by Visual Studio the app runs fine. So one solution (at least a temporary one, but easy to implement ;) ) would be make the automatic adding of the included sqlite3.dll optional. If I recall correctly there was a build definition for this in the past?
This way a developer targeting Windows 10 (or maybe other platforms in the future) could opt-out of the bundled .dll and use their own.

**Comment by ericsink** [comment]
Could you provide more detail about "the sqlite3.dll provided by Visual Studio"?  Exactly where does this come from?  What version of Visual Studio are you using?  etc.

There are two issues here:

(1)  If I can build my sqlite3.dll differently (better), then I want to figure out how.

(2)  I need to re-look at adding a flag so that Windows users can skip using the bundled sqlite3.dll.

--
E

**Comment by dominik-weber** [comment]
Hi Eric,
I'm using VS 2015. The dll comes from the Visual Studio Gallery: https://visualstudiogallery.msdn.microsoft.com/4913e7d5-96c9-4dde-a1a1-69820d615936
On this page you can also find versions for Win8, 8.1, Windows Phone etc. They are maintained by the SQLite developers directly (I think)

https://visualstudiogallery.msdn.microsoft.com/site/search?f%5B0%5D.Type=User&f%5B0%5D.Value=SQLite%20Development%20Team
The ones without OS version numbers are for Windows(Phone) 8.0. The .vsix downloads can be renamed to .zip for extraction.

I'm not too firm on licensing and such but maybe you can just include the prebuilt dlls from these projects?
Anyhow I think adding a skip-bundled-dll flag would be really useful, for example if someone needs a custom built sqlite.dll.

**Comment by dominik-weber** [comment]
Btw. I just realized you can also download the .vsix files directly from the sqlite website https://www.sqlite.org/download.html

**Comment by mjracca** [comment]
Since migrating to Visual Studio 2015, I'm having this issue again as @dominik-weber. 

To fix it, I needed to add this two references:
![image](https://cloud.githubusercontent.com/assets/3976257/10608423/57204ee4-7715-11e5-9b84-71ef0fbbc12e.png)

I just tested it on a desktop, but I'm not sure yet if it will work cross all platforms.

**Comment by dominik-weber** [comment]
did you already have time to look at the prebuilt assemblies?

**Comment by ericsink** [comment]
Not exactly sure what you mean by "the prebuilt assemblies", but I was looking at stuff related to this issue this week.  More to come, hopefully.

**Comment by dominik-weber** [comment]
I'm sorry, I meant the prebuilt sqlite3.dll files you can download from the sqlite website: https://www.sqlite.org/download.html (download .vsix files, open them with zip program)

**Comment by ericsink** [comment]
Ah, yes.

There are two issues here:

(1)  Something about the sqlite3.dll I'm including isn't working in some situations, none of which I have reproduced or understand completely yet.

(2)  Regardless of how I resolve (1), I want (and thought I already had) better support for use cases where the sqlite3.dll for Windows is being provided from Somewhere Else.

And there are two ways to resolve (1):

(a)  Fix the way I build SQLite for whatever environments it's not working in.

(b)  Don't build SQLite.  Just ship the dll Richard and his team builds.

And there are a couple of possible ways to do (a):

(x)  Fix my build scripts.

(y)  Use the exact build script from the sqlite source tree.

I could expound further on (x) and (y), but I'll stop here.  Suffice it to say that lots of possibilities are in play.

But among them, (b) is my very last resort.  It is the option I plan to avoid unless I fail at everything else.

**Comment by dominik-weber** [comment]
Hmm, I don't really see why building the sqlite3.dll yourself would improve anything apart from being more consistent with other platforms where you might not have access to a prebuilt library. In my mind using the one from sqlite.org would reduce the maintenance required to keep the dll up to date, but i guess that just comes down to personal preference.

Regarding (2): Did you add any means of using external sqlite3.dll builds? Currently when adding a reference to sqlitepcl_raw AND the prebuilt vsix/dll i'm unable to build the project: "Payload contains two or more files with the same destination path 'sqlite3.dll'."
What I'm doing to work around this issue is removing the file build\netcore451\SQLitePCL.raw_basic.targets from the nuget cache directory. However this is not a clean solution at all since it has to be done manually on every development machine and also affects other projects (at least if you have a shared nuget cache). One solution to this would be to simply add a compile variable which prevents the bundled sqlite3.dll from being injected.

**Comment by ericsink** [comment]
Sorry, I should have explained a bit more.

Yes, I want to build sqlite myself so that all of the ones I ship are built with the same options.  Also, I want to understand why the one I'm building is having problems, and once I do, there will be no need to use the one from sqlite.org.  In other words, I don't want to ship the ones from sqlite.org just to avoid figuring out what's wrong.

Regarding (2):  Not yet.  Should be simple.  I tried, and ran into a glitch and didn't push through it yet.  Soon.

**Comment by ericsink** [comment]
Could you try version 0.8.5-pre1 (from nuget.org)?

In your project file, set the msbuild property UseSQLiteFrom to the value "elsewhere":

    <UseSQLiteFrom>elsewhere</UseSQLiteFrom>

Hopefully this works, although it only resolves (2).

**Comment by dominik-weber** [comment]
works perfectly, thanks! :)

**Comment by onovotny** [comment]
Might be relevant here but the latest Win10 SDK now seems to include SQLite too:
https://blogs.windows.com/buildingapps/2015/12/03/windows-10-build-10586-sdk-breakdown/
https://msdn.microsoft.com/en-us/library/windows/apps/xaml/mt592864.aspx

Not yet sure how it interacts with this project.

**Comment by ericsink** [comment]
@onovotny Thanks for the heads-up.

**Comment by tuath** [comment]
Any news on this for uwp apps ? is the new recommended way to use your nugget package but for uwp, use external libraries though the vsix ?

**Comment by ericsink** [comment]
"is the new recommended way to use your nugget package but for uwp, use external libraries though the vsix ?"

Yes, that is the current workaround, with:

    <UseSQLiteFrom>elsewhere</UseSQLiteFrom>

The bundled sqlite3.dll in SQLitePCL.raw is still broken for UWP, but I am now pretty sure I know how to fix it.  I hope to get this done next week.  Our company is moving to new office space this week, and that has caused me a variety of delays.  :-)

**Comment by ericsink** [comment]
Clarification:  Using the sqlite3.dll from the vsix (or the Windows SDK) is more than just "the current workaround".  For some people, it is the preferred way of working, and I will continue to support this going forward.

**Comment by tuath** [comment]
Great news ! since I am using xamarin, I would prefer to have just "one way" of using sqlite hence your package, with the fix for uwp... I will be waiting :)
Let me know i can help test it out !
Regards, Erik.

**Comment by dominik-weber** [comment]
in the second link from @onovotny there is actually a section on how to build sqlite yourself, maybe this helps.

**Comment by ericsink** [comment]
TLDR:  Please try SQLitePCL.raw 0.8.5-pre2

I am about 95% certain that all the problems described in the thread above are caused by the inability to find the Visual C++ 2013 runtime DLL.  I just pushed up 0.8.5-pre2, which contains an attempted fix.

In a nutshell, the "fix" is an SDKReference item in the .targets file.  I've tried this before and never quite gotten the magic incantation to work.  This time it would seem that I've got it.  Furthermore, if this fix is correct, then I am a bonehead because I could and should have figured it out a long time ago.

Another way of saying the above is that I am 95% certain there was nothing wrong with the way I was building sqlite3.dll.  It was simply failing because it needed a reference to the Visual C++ runtime DLL.  The SQLite vsix was succeeding because it was automatically adding that reference.  As of 0.8.5-pre2, the nuget package should be doing so as well.

Note that I am *not* (yet) building specifically for UWP.  This change is simply relying on the ability of a UWP app/library to consume a nuget package targeted at netcore451.  Since I am not using any UWP-specific APIs, building specifically for UWP should not be necessary.  I do have some of the groundwork laid for later supporting the "dotnet" TFM and for building with the v140 PlatformToolset, but none of that is used or necessary yet.  I am also going to be exploring the possibilities of using the "runtimes" folder in nuget 3.x, but probably not until that functionality is supported by Xamarin.

Currently, 0.8.5-pre2 works for me when using Windows 8 Universal and when using Windows 10 Universal in Visual Studio 2015.  Feedback from others would be welcomed.

**Comment by ps30** [comment]
Hi Eric, just a FYI Microsoft is ending support for .Net 4.5.1 in January and the new base standard is .Net 4.5.2 moving forward.  Can you adjust this in your next release?
http://blogs.msdn.com/b/dotnet/archive/2014/08/07/moving-to-the-net-framework-4-5-2.aspx

Also, please explain why we need to have an external dependency on the old 2013 Visual C++ run-time libraries being installed on every machine (we are using your code with .Net 4.5.2 WPF C# on Windows 10 64 bit). SQLite3.dll should be compiled to statically link the C++ libs under windows or at a minimum use the better/newer VS2015 compiler correct?

Thanks for all the amazing work - we all really enjoy using your code in our .Net apps under Windows.

**Comment by ericsink** [comment]
(1)  Yeah, good point.  I do need to change that to netcore452

(2)  I am still working on a VC 2015 build.  The build itself is easy.  I just haven't figured out all the new NuGet 3.x issues of getting that build into the nupkg.

(2a)  I *wish* I could statically link the C++ runtime lib, but all information I have found seems to say that Microsoft does not support this anymore.  There is some mention of this further up in this thread.

**Comment by ericsink** [comment]
Slight and ultra-pedantic correction to (1) above:  netcore451 does not precisely refer to .NET 4.5.1.  Rather, it is equivalent to the TFM for Win81.

That said, I do have some updates that should be done w.r.t. the end of support for certain .NET framework versions as mentioned above.

**Comment by tuath** [comment]
Hi Eric, just tried SQLitePCL.raw 0.8.5-pre2 with my uwp app, it works on windows 10 desktop but not when i deploy to windows 10 mobile (x86 using hyper-v emulator)... any ideas ?

**Comment by ericsink** [comment]
Hmmm.  Any error messages or other diagnostic info?

**Comment by tuath** [comment]
nope... just the same "could not load sqlite3.dll"

**Comment by adrianstevens** [comment]
I'm seeing the same issue.  I have a UWP project that I can run locally (i.e. on a desktop) but if run on a Windows 10 mobile simulator or device (i.e. Lumia 640) it throws a runtime error:

"Unable to load DLL 'sqlite3': The specified module could not be found." 

I've reproduced this with v0.8.2 and 0.8.5-pre2.  I'm happy to do further testing or send additional details.

**Comment by ericsink** [comment]
I've been continuing to work on this area.

Current conjecture:  UWP claims to be able to handle stuff compiled for win81 TFM, but maybe this is not necessarily true for native code.  Like maybe it works fine for desktop stuff, but it fails for a device or simulator because it can't just drop the VS 2013 C runtime into that environment and expect it to work.

So maybe, because my nupkg has a mixture of managed code and native code, I need to be more specifically targeted for UWP.  Maybe I need the native code to be compiled against the VC 2015 runtime instead of the 2013 runtime.  Maybe I need to use the uap10 TFM or the dotnet TFM.  

So this is the path I'm currently on.  The actual compile steps have been pretty simple.  Getting stuff into a nupkg that UWP can consume (without breaking older use cases) has been the tough part.  Most days I feel like maybe I'm the first person on the planet trying to create a cross-platform bait and switch PCL nuget package with support for UWP as well as older platforms  and bundling both native and manage code.

**Comment by onovotny** [comment]
I think if you use the new `\ref\uap10.0` and `\runtimes\uap10.0\` support, it'll work because only UWP will read those. Existing NuGet v2 will ignore it and just use `\lib\foo`

**Comment by ericsink** [comment]
@onovotny Thanks for chiming in here.  I thought about using `\ref\` and `\runtimes\`, but my package uses a `.targets` file to allow different builds of the native code to be selected based on an msbuild property.  So I've been trying to put the assembly in `\lib\uap10.0\SQLitePCL.raw.dll` and the targets file in `\build\uap10.0\SQLitePCL.raw.targets`, on the belief that this mechanism would still work as long as I also include the appropriate dependencies/group targetFramework="uap10.0" stuff.  Am I on the wrong track?

**Comment by onovotny** [comment]
I think that if you go down the route of using ref/runtimes for UWP then I'd skip the targets. The targets mechanism was a workaround for NuGet v2 for not supporting those runtime pivots. You could have a `\build\uap10.0\_._` folder to explicitly not include the targets for `uap10.0` as each top level folder evaluates independently. Having a blank uap10.0 build folder would override that for just that TFM.

**Comment by ericsink** [comment]
I'm (currently) using the targets file for more than just choosing CPU arch.  The `UseSQLiteFrom` msbuild property can be used to say "don't provide me with sqlite3.dll at all, since I want to use the one from the vsix built by the sqlite team".  OR, that same msbuild property can be used to say "instead of regular sqlite3.dll, provide me with SQLCipher".

Originally, my plan was to separate these use cases into separate nuget packages.  So for example I would have one form of the SQLitePCL.raw nupkg which bundled plain SQLite, one which bundled SQLCipher, and one which bundled nothing.

This approach caused big problems for library packages like Akavache, which build on SQLitePCL.raw.  The decision of whether to include the encryption features (SQLCipher) or not should be left up to the app developer, but Akavache doesn't want to maintain N versions simply because I have N versions.

At the time, introducing the msbuild property was the easy choice because I already had to swallow the pain of that darn targets file anyway just to get CPU arch selection.  As of NuGet v3, the CPU arch stuff could be done far more elegantly by the ref and runtimes folders, but AFAICT, a targets file is still the only way to people to use msbuild properties to customize which things are provided.

(Furthermore, I remind myself here that I couldn't really switch everything to the more elegant ref and runtimes folders until they are supported by Xamarin anyway.)

Typically this line of reasoning ends with me wishing that NuGet had something like Debian's "virtual packages", so I could say "the essence of SQLitePCL.raw is provided by any of the following concrete packages..."  But of course, that would be a large and entirely new can of worms.

**Comment by onovotny** [comment]
Fair enough - afaik, there's nothing that shouldn't continue to work with `uap10.0` with the targets file. It'd be similar to what you've done with WP8... 

For the dependencies, looking at your build process as you appear to generate the nuspec file, if you want to use my RefGen tool, you'll prob need to invoke it manually instead of relying on its targets.

https://github.com/onovotny/ReferenceGenerator#command-line

For arg 0, you'll need to find out the value of `$(NuGetTargetMoniker);$(_NuGetTargetFallbackMoniker)` from your UWP project and use that.

**Comment by ericsink** [comment]
Update, FWIW:  I have reproduced the behavior others are seeing, where it works for desktop but not in the mobile simulator.

Still digging...

**Comment by ericsink** [comment]
@onovotny Yeah, my gen_build.cs is probably unique.  I just hate maintaining dozens of XML files that are similar.  Generating them gives me a way to change something in just one place.  I borrowed the idea from CMake.

Anyway...

Looks like using NuSpec.ReferenceGenerator will be fairly painless.  I fixed gen_build.cs to add it to the appropriate project.json file(s), and on first glance it looks like it Just Works.  I've got a bit more work to do, but right now this seems like the way to go.  Thanks.

**Comment by onovotny** [comment]
Awesome! I tried to have it do a lot via convention to find sensible defaults. The one parameter that might need updating, if any, is the `NuSpecFile` if it can't find it. By default it looks for a `.nuspec` file at or beneath the sln file that matches the project name. If the nuspec name differs, then it'd need to be specified. However, if you're seeing the `uap10.0` dependencies group, then it's clearly working for you :)

**Comment by onovotny** [comment]
Looking at the checkin comment, it looks like you need to update both _raw and _basic nuspecs? If that's the case, you could probably write/generate a target that's run after `UpdateNuspecDependencies` that calls RefGen again with another nuspec. Alternatively, I'd be happy to accept a PR that enables arg 2 to accept a semi-colon joined list of nuspecs and update all of them in one pass:
https://github.com/onovotny/ReferenceGenerator/blob/master/src/ReferenceGenerator/Program.cs#L37

**Comment by ericsink** [comment]
Cool.  I'm not sure yet how I will deal with _basic.nuspec, but thanks for the tip.

**Comment by ericsink** [comment]
OK.  Version 0.8.5-pre3 is in nuget and contains support for uap10.0.  Limited testing indicates that it works for a universal windows app in both desktop and mobile simulator.  Additional feedback and testing would be welcomed.  :-)

**Comment by tuath** [comment]
i'll test it right away !

**Comment by onovotny** [comment]
Dumb question... can you/do you want to support .NET Core across platforms? The managed part would be essentially like the UWP version but you'd need to leverage the NuGet v3 runtimes bits to get the proper runtime/native pivoting across the various Unix flavors/archs (as well as compile SQLite for those)...of course a complex targets file might work too.

**Comment by ericsink** [comment]
Yes, I do think I will want to add support for .NET Core across platforms, but not immediately.  So far I have only climbed just enough of the NuGet v3 learning curve to get UWP support working.

**Comment by tuath** [comment]
Gonna take much longer to test than I thought... nugget in uwp projects behaves differently, and since my app uses SQLite-net-pcl which is still binded to _rawbasic... it insists on restoring it and then I have a version conflict (0.7.1 and 0.8.4-pre3)...

**Comment by tuath** [comment]
ok, so I created a blog post with the description of my findings... http://www.modelon.net/2016/01/nuget-and-project-json/

basically I think I have three options :
1 - understand how to tell nuget for uwp to not restore dependencies of SQLite-net-pcl (which is SQLite-pcl.rawbasic)
2 - For this project, both packages (raw and rawbasic) keep on getting published so that the dependencies for SQLite-net-pcl keep on working
3 - SQLite-net-pcl is upgraded to have a dependency on "raw" and not "rawbasic"

I think option 3 is the one we are looking for ? will create an issue on the other site...

**Comment by ericsink** [comment]
Hmmm.  I should probably continue keeping _basic up-to-date for now.  I should endeavor to get the consuming packages to switch to the new package ID *before* I stop updating the old one.  :-)

**Comment by ericsink** [comment]
OK, I pushed 0.8.5-pre4 up to nuget.  In this one, the _basic form of the package was updated with the uap10.0 stuff as well.

**Comment by tuath** [comment]
Hi Eric, I just tried with the lastest raw_basic and it works perfectly in the emulator with a UWP app.
When I am a bit more advanced in development, will try to deploy on a real device (although this item is fixed, there are plenty bugs remaining in xamarin forms for uwp...)

Good job !

**Comment by tuath** [comment]
just verified, it also works great when deployed to a real device !

**Comment by ericsink** [comment]
Yay.  After reviewing a bit further to make sure I didn't miss anything, I'll close this issue after my next push of a non-pre nuget package.

**Comment by ericsink** [comment]
0.8.5, with uap10.0 support, has been pushed to nuget

**Comment by adrianstevens** [comment]
Thanks Eric - fantastic

**Comment by ghuntley** [comment]
Thank-you Eric!

**Comment by ghuntley** [comment]
UWP + Akavache works, make sure you upgrade to latest SQLite from last nighjt and read this https://github.com/ghuntley/repros/tree/master/nuget/TransitiveRefs to work around a nuget resolution issue.

**Comment by ps30** [comment]
Eric great work!  Just wanted to be clear on the requirement for Microsoft C++ libraries and SQLitePCL.raw.  Can you please make clear here and in docs, 1: if SQLitePCL.raw 0.8.5 is compiled using the current Visual Studio 2015 and 2: are the C++ redist libraries statically linked or are we required to have to install the Microsoft 2013  - or hopefully the better 2015 32/64 bit libs on every machine that we intend to use SQLitePCL.raw on.  Sqlite.org does distribute the compiled form of sqlite.dll statically linked so all you need is the dll in your application running on Windows.

I know this applies mostly to Windows 8.X and 10 32/64bit machines but it is important to know to use SQLitePCL.raw correctly.  Also Win10 does not install or include most of the older redist libs and this must be done by the app dev to get things working.

I am sure others would like to know the answers to these two important questions so it might be good to add a line in the docs so people know when deploying using SQLitePCL.raw on Windows.

**Comment by ericsink** [comment]
Yes, this belongs in "the docs", but for now, I'll answer here:

SQLitePCL.raw 0.8.5 contains several copies of the sqlite3 native code in
compiled form, each compiled with a different version of the Visual C++
platform toolset and linked against the corresponding C runtime libraries.

These include:

v110
v110_xp
v110_wp80
v120
v120_wp81
v140

thus supporting everything from Windows XP through Windows 10 and Windows
10 Mobile.

In general, this is setup so that you'll hopefully get what you want for a
given nuget platform without extra hassle.

For example, if you are building a UWP application, nuget should select
uap10.0, which will bring in the native code built with v140, which will be
linked against the Visual C++ 2015 C runtime libs.

As I've said before (read through this entire thread from the top), it is
my understanding that Microsoft no longer supports statically linking the C
runtime as of VS 2013 and 2015.  If I am wrong about this, I will be happy
to stand corrected and fix things.  Static linking against the C runtime
would be my preferred solution.

But on the assumption that I am correct, I believe you are incorrect when
you say "Sqlite.org does distribute the compiled form of sqlite.dll
statically linked so all you need is the dll in your application running on
Windows.".  Or rather, you may be correct for some versions of Windows, but
not all.  For example, maybe they are compiling with a v110 toolset, which
does support static linking against libcrt.  But this, as I understand it,
might work for any desktop version of Windows, but would never work for a
UWP app.

To be clear (and redundant):  I know you want me to statically link the C
runtime and have everything Just Work on every Windows platform without
needing to install a C runtime separately.  I want that too.  But it's
impossible.  If I'm wrong about that, show me.

Further detail:

For Windows 8.1, SQLitePCL.raw 0.8.5 injects an SDKReference for the Visual
C++ 2013 C runtime library, as long as the current build is not "Any CPU".
This helps prevent people from forgetting to add the VC++ extension SDK to
their build.

Currently, no such SDKReference is generated for the Visual C++ 2015
library, because testing indicates it to not be necessary.  I assume this
is because I haven't run across a Windows 10 platform (desktop or mobile)
which did not already have the necessary CRT DLL pre-installed.

Nothing I've written here specifically deals with issues of installation
and deployment.  I'm talking merely about what CRT I am linking against,
and I'm making hand-wavy remarks about whether that version of the CRT
might be expected to already be installed on certain versions of Windows.

Looking over what I have written here so far, I'm not sure this is the
"clear" answer you were hoping for.  :-)  But I *think* this is because
your question contains implicit assumptions that things are somewhat
simpler than they are.

Here's an attempt at a more concise answer:

If you reference the nuget package in such a manner that you get the
"win81" stuff, then you will get a version of the sqlite native bits that
are dynamically linked against the VC++ 2013 C runtime libraries, and if
those are not already on your target system, then you would need to install
them.

Same for "uap10.0" -- you'll need the VC++ 2015 C runtime libs.

Finally, let me say that I am currently working on some improvements in
this area, but I don't expect to find a magic bullet solution.

I hope this helps.  Feedback welcome.

--
E
